File: src/mm/pmm.c
/* Physical page allocator: static bitmap over the Limine memory map.
 *
 * bit = 1 -> free, 0 -> used/unusable. Only Limine "usable" entries become
 * free; kernel image, modules, reclaimable and firmware regions stay used.
 */
#include "mm/pmm.h"
#include "kprintf.h"
#include "lib/string.h"

#define PMM_MAX_PHYS (512ULL << 20)                 /* cap: 512 MiB */
#define BITMAP_PAGES (PMM_MAX_PHYS / PAGE_SIZE)     /* 131072 pages */
#define BITMAP_WORDS (BITMAP_PAGES / 64)            /* 2048 words = 16KiB */

static uint64_t bitmap[BITMAP_WORDS];  /* starts zeroed: everything used */

static size_t total_pages, free_pages;

/* --- Limine requests ------------------------------------------------------- */

struct limine_memmap_entry {
    uint64_t base, length, type;
};
#define LIMINE_MEMMAP_USABLE 0

struct limine_memmap_response {
    uint64_t revision, entry_count;
    struct limine_memmap_entry **entries;
};
struct limine_hhdm_response { uint64_t revision, offset; };

struct limine_memmap_request {
    uint64_t id[4];
    uint64_t revision;
    struct limine_memmap_response *response;
};
struct limine_hhdm_request {
    uint64_t id[4];
    uint64_t revision;
    struct limine_hhdm_response *response;
};

__attribute__((used, section(".limine_requests")))
static volatile struct limine_memmap_request memmap_req = {
    .id = { 0xc7b1dd30df4c8b88ULL, 0x0a82e883a194f07bULL,
            0x67cf3d9d378a806fULL, 0xe304acdfc50c3c62ULL },
};
__attribute__((used, section(".limine_requests")))
static volatile struct limine_hhdm_request hhdm_req = {
    .id = { 0xc7b1dd30df4c8b88ULL, 0x0a82e883a194f07bULL,
            0x48dcf1cb8ad2b852ULL, 0x63984e959a98244bULL },
};

uint64_t hhdm_offset;

/* --- bitmap ops ------------------------------------------------------------- */

static inline void bm_set(size_t i)   { bitmap[i >> 6] |=  (1ULL << (i & 63)); }
static inline void bm_clear(size_t i) { bitmap[i >> 6] &= ~(1ULL << (i & 63)); }
static inline int  bm_test(size_t i)  { return (bitmap[i >> 6] >> (i & 63)) & 1; }

void pmm_init(void) {
    if (!memmap_req.response || !hhdm_req.response) {
        KLOG_ERR("pmm: missing limine memmap/hhdm response");
        for (;;) __asm__ volatile ("hlt");
    }
    hhdm_offset = hhdm_req.response->offset;
    KLOG_INFO("pmm: hhdm offset=%p", (void *)hhdm_offset);

    struct limine_memmap_response *mm = memmap_req.response;
    uint64_t top = 0;

    for (uint64_t i = 0; i < mm->entry_count; i++) {
        struct limine_memmap_entry *e = mm->entries[i];
        if (e->type != LIMINE_MEMMAP_USABLE)
            continue;

        uint64_t base = (e->base + PAGE_SIZE - 1) & ~(PAGE_SIZE - 1);
        uint64_t end  = (e->base + e->length) & ~(PAGE_SIZE - 1);
        if (end > PMM_MAX_PHYS) end = PMM_MAX_PHYS;
        if (base >= end) continue;

        for (uint64_t p = base; p < end; p += PAGE_SIZE) {
            bm_set(p / PAGE_SIZE);
            free_pages++;
            total_pages++;
        }
        if (end > top) top = end;
    }
    KLOG_INFO("pmm: %lu pages free (%lu MiB), top=%#lx",
              free_pages, free_pages * PAGE_SIZE >> 20, top);
}

uintptr_t pmm_alloc_page(void) {
    for (size_t w = 0; w < BITMAP_WORDS; w++) {
        if (!bitmap[w]) continue;
        int bit = __builtin_ctzll(bitmap[w]);
        size_t idx = (w << 6) + (size_t)bit;
        bm_clear(idx);
        free_pages--;
        return (uintptr_t)idx * PAGE_SIZE;
    }
    return 0;
}

uintptr_t pmm_alloc_pages(size_t count, size_t align) {
    if (align < PAGE_SIZE) align = PAGE_SIZE;
    size_t align_pages = align / PAGE_SIZE;

    for (size_t i = 0; i + count <= BITMAP_PAGES; i++) {
        if (i % align_pages) { i -= i % align_pages; continue; }
        size_t j = 0;
        for (; j < count; j++)
            if (!bm_test(i + j)) break;
        if (j == count) {
            for (j = 0; j < count; j++) bm_clear(i + j);
            free_pages -= count;
            return (uintptr_t)i * PAGE_SIZE;
        }
        i += j;   /* skip past the used page (loop adds 1) */
    }
    return 0;
}

void pmm_free_page(uintptr_t pa) {
    size_t idx = pa / PAGE_SIZE;
    if (pa % PAGE_SIZE || idx >= BITMAP_PAGES || bm_test(idx)) {
        KLOG_ERR("pmm: bad free of %p", (void *)pa);
        return;
    }
    bm_set(idx);
    free_pages++;
}

size_t pmm_free_count(void)  { return free_pages; }
size_t pmm_total_count(void) { return total_pages; }

/* --- test ------------------------------------------------------------------- */

void pmm_selftest(void) {
    size_t before = pmm_free_count();

    uintptr_t a = pmm_alloc_page();
    uintptr_t b = pmm_alloc_page();
    if (!a || !b || a == b || (a % PAGE_SIZE) || (b % PAGE_SIZE))
        KLOG_ERR("pmm test FAIL: a=%#lx b=%#lx", a, b);
    if (pmm_free_count() != before - 2)
        KLOG_ERR("pmm test FAIL: count after alloc = %lu", pmm_free_count());

    pmm_free_page(a);
    if (pmm_free_count() != before - 1)
        KLOG_ERR("pmm test FAIL: count after one free");

    uintptr_t run = pmm_alloc_pages(4, PAGE_SIZE);
    if (!run || run % PAGE_SIZE)
        KLOG_ERR("pmm test FAIL: 4-page run");
    pmm_free_page(b);
    KLOG_INFO("pmm: selftest ok (a=%#lx b=%#lx run=%#lx)", a, b, run);
}

File: kernel-rs/src/mm/pmm.rs
//! Physical page allocator: static bitmap over the Limine memory map.
//!
//! bit = 1 -> free, 0 -> used/unusable. Only Limine "usable" entries become
//! free; kernel image, modules, reclaimable and firmware regions stay used.

use core::cell::UnsafeCell;

use crate::limine;

pub const PAGE_SIZE: u64 = 4096;

const PMM_MAX_PHYS: u64 = 512 << 20; // cap: 512 MiB
const BITMAP_PAGES: usize = (PMM_MAX_PHYS / PAGE_SIZE) as usize; // 131072 pages
const BITMAP_WORDS: usize = BITMAP_PAGES / 64; // 2048 words = 16KiB

struct BitmapCell(UnsafeCell<[u64; BITMAP_WORDS]>);
unsafe impl Sync for BitmapCell {}

/// Starts zeroed: everything used.
static BITMAP: BitmapCell = BitmapCell(UnsafeCell::new([0; BITMAP_WORDS]));

static mut TOTAL_PAGES: usize = 0;
static mut FREE_PAGES: usize = 0;

/// Higher-half direct map base, provided by Limine.
pub static mut HHDM_OFFSET: u64 = 0;

#[inline(always)]
pub fn pa_to_va(pa: u64) -> *mut u8 {
    unsafe { (pa + HHDM_OFFSET) as *mut u8 }
}

#[inline(always)]
pub fn va_to_pa(va: u64) -> u64 {
    unsafe { va - HHDM_OFFSET }
}

fn bm_set(i: usize) {
    unsafe { (*BITMAP.0.get())[i >> 6] |= 1 << (i & 63) };
}

fn bm_clear(i: usize) {
    unsafe { (*BITMAP.0.get())[i >> 6] &= !(1 << (i & 63)) };
}

fn bm_test(i: usize) -> bool {
    unsafe { ((*BITMAP.0.get())[i >> 6] >> (i & 63)) & 1 != 0 }
}

pub fn init() {
    let (mm, hhdm) = match (limine::memmap_response(), limine::hhdm_response()) {
        (Some(mm), Some(h)) => (mm, h),
        _ => {
            crate::klog_err!("pmm: missing limine memmap/hhdm response");
            loop {
                crate::io::hlt();
            }
        }
    };

    unsafe {
        HHDM_OFFSET = hhdm.offset;
        crate::klog_info!("pmm: hhdm offset={:#x}", HHDM_OFFSET);
    }

    let mut top: u64 = 0;

    for i in 0..mm.entry_count as usize {
        let e = unsafe { &*mm.entries.add(i).read() };
        if e.typ != limine::LIMINE_MEMMAP_USABLE {
            continue;
        }

        let base = (e.base + PAGE_SIZE - 1) & !(PAGE_SIZE - 1);
        let mut end = (e.base + e.length) & !(PAGE_SIZE - 1);
        if end > PMM_MAX_PHYS {
            end = PMM_MAX_PHYS;
        }
        if base >= end {
            continue;
        }

        let mut p = base;
        while p < end {
            bm_set((p / PAGE_SIZE) as usize);
            unsafe {
                FREE_PAGES += 1;
                TOTAL_PAGES += 1;
            }
            p += PAGE_SIZE;
        }
        if end > top {
            top = end;
        }
    }
    unsafe {
        crate::klog_info!(
            "pmm: {} pages free ({} MiB), top={:#x}",
            FREE_PAGES,
            FREE_PAGES as u64 * PAGE_SIZE >> 20,
            top
        );
    }
}

pub fn alloc_page() -> u64 {
    for w in 0..BITMAP_WORDS {
        let word = unsafe { (*BITMAP.0.get())[w] };
        if word == 0 {
            continue;
        }
        let bit = word.trailing_zeros() as usize;
        let idx = (w << 6) + bit;
        bm_clear(idx);
        unsafe {
            FREE_PAGES -= 1;
        }
        return idx as u64 * PAGE_SIZE;
    }
    0
}

pub fn alloc_pages(count: usize, align: usize) -> u64 {
    let align = align.max(PAGE_SIZE as usize);
    let align_pages = align / PAGE_SIZE as usize;

    let mut i = 0usize;
    while i + count <= BITMAP_PAGES {
        if i % align_pages != 0 {
            i -= i % align_pages;
            i += 1; // C for-loop increment on `continue`
            continue;
        }
        let mut j = 0;
        while j < count {
            if !bm_test(i + j) {
                break;
            }
            j += 1;
        }
        if j == count {
            for k in 0..count {
                bm_clear(i + k);
            }
            unsafe {
                FREE_PAGES -= count;
            }
            return i as u64 * PAGE_SIZE;
        }
        i += j; // skip past the used page (loop adds 1)
        i += 1;
    }
    0
}

pub fn free_page(pa: u64) {
    let idx = (pa / PAGE_SIZE) as usize;
    if pa % PAGE_SIZE != 0 || idx >= BITMAP_PAGES || bm_test(idx) {
        crate::klog_err!("pmm: bad free of {:#x}", pa);
        return;
    }
    bm_set(idx);
    unsafe {
        FREE_PAGES += 1;
    }
}

pub fn free_count() -> usize {
    unsafe { FREE_PAGES }
}

pub fn total_count() -> usize {
    unsafe { TOTAL_PAGES }
}

pub fn selftest() {
    let before = free_count();

    let a = alloc_page();
    let b = alloc_page();
    if a == 0 || b == 0 || a == b || a % PAGE_SIZE != 0 || b % PAGE_SIZE != 0 {
        crate::klog_err!("pmm test FAIL: a={:#x} b={:#x}", a, b);
    }
    if free_count() != before - 2 {
        crate::klog_err!("pmm test FAIL: count after alloc = {}", free_count());
    }

    free_page(a);
    if free_count() != before - 1 {
        crate::klog_err!("pmm test FAIL: count after one free");
    }

    let run = alloc_pages(4, PAGE_SIZE as usize);
    if run == 0 || run % PAGE_SIZE != 0 {
        crate::klog_err!("pmm test FAIL: 4-page run");
    }
    free_page(b);
    crate::klog_info!("pmm: selftest ok (a={:#x} b={:#x} run={:#x})", a, b, run);
}
